drafts/bitcartel-nu0-tx-format/draft.rst

```diff
@@ -0,0 +1,202 @@
+
+Existing code typically checks the tx version using greater than comparison operators::
+
+    if (tx.nVersion >= 2) {
+      for (int js = 0; js < joinsplits; js++) {
+
+Existing tests typically set tx.nVersion to zero as an error condition::
+
+    mtx.nVersion = 0;
+    // https://github.com/zcash/zcash/blob/59de56eeca6f9f6f7dc1841630d53676075242a5/src/gtest/test_mempool.cpp#L99
+
+    EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-txns-version-too-low", false)).Times(1);
+    // https://github.com/zcash/zcash/blob/30d3d2dfd438a20167ddbe5ed2027d465cbec2f0/src/gtest/test_checktransaction.cpp#L99
+
+By using a negative value for the version field, we ensure there is replay protection between Legacy and Overwinter compatible software.
+
+Consider an example where the raw little-endian bytes of the version field of an Overwinter transaction begin 0xFDFFFFFF ( binary 11111101111111111111111111111111 )
+
+Legacy parsers will deserialize the raw version field as a 32-bit signed integer.  With a negative version value of -3, legacy parsers will reject the transaction.
+
+Overwinter parsers will deserialize the raw version field as a 32-bit signed integer.  With a negative raw value of -3, the Overwinter parser will accept the transaction as the most significan bit of the 32-bit signed integer has been set.
+
+Overwinter parsers can retrieve the transaction format version of 3 by getting the absolute value of the raw version field e.g. using standard library call std::abs()
+
+Currently, the nVersion field is a public member variable which can be accessed directly.  As part of implementing Overwinter, the nVersion field will be made private with access restricted to using getters, e.g.::
+
+    bool isLegacyFormat()         // return true if the msb of nVersion is not set
+    unsigned int32 getVersion()   // return absolute value of raw version field which is compatible with Legacy and Overwinter
+
+
+Forwards Compatibility
+----------------------
+
+A branch may support many transaction version formats.  For example:
+
+* Zcash reference implementation, branch "Zcash", versions 3, 4.
+* Fork of Zcash, branch "Clone", versions 3, 4*
+
+Where transaction format version 4* for the "Clone" branch might be substantially different from the expected transaction format version 4 for the "Zcash" branch.
+
+Given forwards compatibility, we want the "Zcash" branch nodes to accept transaction version 4, whilst rejecting version 4* transactions which are intended only for the "Clone" branch.
+
+To achieve this, Overwinter requires a transaction to include a branch ID, to explicitly state which branch of the network this transaction is intended for.
+
+Overwinter introduces a new signature hashing scheme which includes the branch ID, but by including the branch ID into the transaction format, clients can quickly reject transactions during deserialization without having to check signatures.
```

Comment:
We decided against this (and to wear the corresponding verification overhead), to prevent third-party parsers being forced to update after Sapling (and other network upgrades) to maintain support for non-deprecated transaction formats.

The ZIP should instead explain that the branch ID in the transaction format is for domain separation of `nVersion`, so that the combination `nVersion | branchID` can be assumed unique across all branches. We only need to enforce this at parsing and serialization time; everywhere else, we can just rely on `nVersion` (as mentioned above).

Reply:
Hmm, a bit unclear.  Could you clarify?

`nVersion | branchID` is this bytes [nVersion] followed by bytes [branchID] , or is this a value computed by `nVersion OR branchID` ?

`We only need to enforce this at parsing and serialization time` - what does enforce mean?  So when parsing, 
- (i) signatures are still checked against the branch ID
- (ii) tx is rejected if the branch ID (but if this can occur earlier during the parsing stage than (i) why not?)

Reply:
Sorry, I should have said `nVersion || branchID` - ie. the first eight bytes of the transaction uniquely specify all remaining bytes of the transaction format.

> what does enforce mean?

I mean that because every `nVersion` is uniquely associated with a branch ID from our (main) chain, we can use `nVersion || branchID` to reject transaction formats from outside our chain, and (by convention) ensure uniqueness of `nVersion` across transaction formats inside our chain. Thus in-memory (after transactions have been parsed, or before they have been serialized), we can just rely on `nVersion / getVersion()` for referring to a transaction format version, rather than needing to check `nVersion || branchID` every time.

> So when parsing,
> - (i) signatures are still checked against the branch ID

No - this happens after parsing, when checking transaction validity.

> - (ii) tx is rejected if the branch ID (but if this can occur earlier during the parsing stage than (i) why not?)

This is *consensus* rejection, not *parser* rejection. At this point, the node is enforcing two-way replay protection, so that transactions with formats we know how to parse (e.g. Overwinter v3 transactions created after the Sapling upgrade from anyone that doesn't follow Sapling) won't be considered valid on our main chain.

Reply:
To be clear, what we previously considered (and decided against) was using `branchId` for both parser rejection *and* consensus rejection. But even if we were doing that, we'd still be doing those as separate steps in separate parts of the code.